README.md
fc-java-sdk
=======

[![maven version][mvn-image]][mvn-url]
[![build status][travis-image]][travis-url]

[mvn-image]: https://img.shields.io/maven-central/v/com.aliyun/aliyun-java-sdk-fc.svg?style=flat-square
[mvn-url]: http://search.maven.org/#search%7Cga%7C1%7Caliyun-java-sdk-fc
[travis-image]: https://img.shields.io/travis/aliyun/fc-java-sdk/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/aliyun/fc-java-sdk.svg?branch=master

## Requirements

- `Java 1.6 and above`

## License

[MIT](LICENSE)

## Install

Add Maven dependencies into pom.xml

```xml
<dependency>
    <groupId>com.aliyun</groupId>
    <artifactId>aliyun-java-sdk-fc</artifactId>
    <version>1.8.24</version>
</dependency>
```

## Example

Create the code directory and write counter nodejs code

```bash
mkdir /tmp/fc_code
cat <<EOF > /tmp/fc_code/counter.js
'use strict';

var counter = 0;
exports.initializer = function(ctx, callback) {
    ++counter;
    callback(null, "");
};

exports.handler = function(event, context, callback) {
  console.log('counter is %d', counter);
  callback(null, String(counter));
};
EOF
```

Run below with your own ENDPOINT, ACCESS_KEY/SECRET_KEY and ACCOUNT_ID environment variables

```Java
import com.aliyuncs.fc.client.FunctionComputeClient;
import com.aliyuncs.fc.constants.Const;
import com.aliyuncs.fc.model.Code;
import com.aliyuncs.fc.request.*;
import com.aliyuncs.fc.response.*;

import java.io.IOException;
import java.net.HttpURLConnection;

public class FcSample {
    private static final String CODE_DIR = "/tmp/fc_code";
    private static final String REGION = "cn-shanghai";
    private static final String SERVICE_NAME = "test_service";
    private static final String FUNCTION_NAME = "test_function";

    public static void main(final String[] args) throws IOException {
        String accessKey = System.getenv("ACCESS_KEY");
        String accessSecretKey = System.getenv("SECRET_KEY");
        String accountId = System.getenv("ACCOUNT_ID");
        String role = System.getenv("ROLE");

        // Initialize FC client
        FunctionComputeClient fcClient = new FunctionComputeClient(REGION, accountId, accessKey, accessSecretKey);

        // Set to a specific endpoint in case needed, endpoint sample: http://123456.cn-hangzhou.fc.aliyuncs.com
        // fcClient.setEndpoint("http://{accountId}.{regionId}.fc.aliyuncs.com.");

        // Create a service
        CreateServiceRequest csReq = new CreateServiceRequest();
        csReq.setServiceName(SERVICE_NAME);
        csReq.setDescription("FC test service");
        csReq.setRole(role);
        CreateServiceResponse csResp = fcClient.createService(csReq);
        System.out.println("Created service, request ID " + csResp.getRequestId());

        // Create a function
        CreateFunctionRequest cfReq = new CreateFunctionRequest(SERVICE_NAME);
        cfReq.setFunctionName(FUNCTION_NAME);
        cfReq.setDescription("Function for test");
        cfReq.setMemorySize(128);
        cfReq.setRuntime("nodejs4.4");
        cfReq.setHandler("counter.handler");
        // Used in initializer situations.
        cfReq.setInitializer("counter.initializer");
        Code code = new Code().setDir(CODE_DIR);
        cfReq.setCode(code);
        cfReq.setTimeout(10);
        CreateFunctionResponse cfResp = fcClient.createFunction(cfReq);
        System.out.println("Created function, request ID " + cfResp.getRequestId());

        // Invoke the function with a string as function event parameter, Sync mode
        InvokeFunctionRequest invkReq = new InvokeFunctionRequest(SERVICE_NAME, FUNCTION_NAME);
        String payload = "Hello FunctionCompute!";
        invkReq.setPayload(payload.getBytes());
        InvokeFunctionResponse invkResp = fcClient.invokeFunction(invkReq);
        System.out.println(new String(invkResp.getContent()));

        // Invoke the function, Async mode
        invkReq.setInvocationType(Const.INVOCATION_TYPE_ASYNC);
        invkResp = fcClient.invokeFunction(invkReq);
        if (HttpURLConnection.HTTP_ACCEPTED == invkResp.getStatus()) {
            System.out.println("Async invocation has been queued for execution, request ID: " + invkResp.getRequestId());
        } else {
            System.out.println("Async invocation was not accepted");
        }
        // Delete the function
        DeleteFunctionRequest dfReq = new DeleteFunctionRequest(SERVICE_NAME, FUNCTION_NAME);
        DeleteFunctionResponse dfResp = fcClient.deleteFunction(dfReq);
        System.out.println("Deleted function, request ID " + dfResp.getRequestId());

        // Delete the service
        DeleteServiceRequest dsReq = new DeleteServiceRequest(SERVICE_NAME);
        DeleteServiceResponse dsResp = fcClient.deleteService(dsReq);
        System.out.println("Deleted service, request ID " + dsResp.getRequestId());
    }
}
```

## API Spec

See: https://help.aliyun.com/document_detail/52877.html


History.md
1.8.35 / 2022-03-21
==================
  * add HTTPInvokeFunctionV2Request for fcapp.run
  * NOTE: this version will not be released to public

1.8.31 / 2021-12-29
==================
  * add UrlInternet&UrlIntranet for http trigger meta

1.8.23 / 2021-10-08
==================
  * add logBeginRule for logConfig

1.8.21 / 2021-06-17
==================
  * add acree instanceID for customContainerConfig

1.8.18 / 2021-04-19
==================
  * add instance metrics support

1.8.15 / 2021-01-09
==================

  * add targetTrackingPolicy for provision

1.8.14 / 2021-01-07
==================

  * add scheduledAction for provision

1.8.12 / 2020-12-16
==================

  * add request metrics support

1.8.11 / 2020-10-27
==================

  * add tracing support

1.8.10 / 2020-09-14
==================

  * add asyncConfig support
  
1.8.8 / 2020-08-25
==================

  * add config validator

1.8.7 / 2020-08-05
==================

  * add getCaPort and getCustomContainerConfig in GetFunctionResponse


1.8.6 / 2020-07-27
==================

  * add connection request timeout

1.8.5 / 2020-07-22
==================

  * add connection request timeout

1.8.4 / 2020-06-22
==================

  * add vpc binding api

1.8.2 / 2020-06-19
==================

  * add customContainerConfig/caPort and instanceType

1.8.1 / 2020-03-16
==================

  * add http method PATCH

1.8.0 / 2020-03-13
==================

  * add SignURL for http trigger

1.7.1 / 2019-12-27
==================

  * Add new field for prepaid list interface
  
1.7.0 / 2019-11-22
==================

  * add async client, for now only support functions and services api
  
1.6.0 / 2019-11-15
==================

  * add on-demand-config API (#87)
  * add instanceConcurrency for Create/Update function (#83)
  * fix tag/untag response without body, requestId already in header

1.5.0 / 2019-09-04
==================
  * Add tag support

1.4.0 / 2019-08-26
==================
  * Add provision interface

1.3.4 / 2019-05-30
==================
  * Set all response encode with utf-8

1.3.3 / 2019-05-23
==================
  * Add Custom Domain HTTPS Support

1.3.2 / 2019-05-23
==================
  * Add Custom Domain HTTPS Support

1.3.1 / 2018-11-22
==================
  * add Get Account Setting

1.3.0 / 2018-11-21
==================

  * add Initializer support
  * add Versioning support 

1.2.1 / 2018-09-20
==================

  * Add Custom Domain Support

1.2.0 / 2018-08-27
==================

  * Add NAS Support

1.1.10 / 2018-06-01
==================

  * Add vpc and cdn events trigger support

1.1.8 / 2018-05-04
==================

  * Fix url encode related problem

1.1.7 / 2018-04-27
==================

  * Add http trigger support

1.1.6 / 2018-04-12
==================

  * Add environment variables support

1.1.5 / 2018-03-16
==================

  * Fixed a compilation problem for jdk 1.9

1.1.4 / 2018-03-09
==================

  * Add time trigger support

1.1.3 / 2017-10-12
==================

  * Fixed FC endpoints for newer regions

1.1.2 / 2017-09-29
==================

  * Refactor log trigger config


src/main/java/com/aliyuncs/fc/config/Config.java
package com.aliyuncs.fc.config;

import com.aliyuncs.auth.AlibabaCloudCredentials;
import com.aliyuncs.auth.AlibabaCloudCredentialsProvider;
import com.aliyuncs.auth.BasicSessionCredentials;
import com.aliyuncs.fc.constants.Const;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

/**
 * TODO: add javadoc
 */
public class Config {

    private static final String ENDPOINT_FMT = "%s.%s.fc.aliyuncs.com";
    private static final String PROPERTIES_FILE = "fc.properties";
    private static final String REGION_PATTERN = "^[A-Za-z0-9\\-\\_]+$";
    private static final String UID_PATTERN = "^[A-Za-z0-9\\-\\_]+$";

    private String endpoint;
    private String apiVersion = "2016-08-15";
    private String accessKeyID;
    private String accessKeySecret;
    private String securityToken;
    private Boolean isDebug = false;
    private String accountId;
    private String uid;
    private AlibabaCloudCredentialsProvider credsProvider = null;

    private int connectTimeoutMillis = Const.CONNECT_TIMEOUT;
    private int connectionRequestTimeoutMillis = Const.CONNECT_TIMEOUT;
    private int readTimeoutMillis = Const.READ_TIMEOUT;
    private int threadCount = 1;
    private int maxConnectCount = 20;
    private int maxPerRoute = 2;

    private String host;
    private String userAgent;

    private Config(String region, String uid, boolean isHttps) {
        checkParam(region, uid);
        this.endpoint = buildEndpoint(region, uid, isHttps);
        this.uid = uid;
        this.userAgent = "";
        try {
            Properties props = new Properties();
            InputStream input = getClass().getClassLoader().getResourceAsStream(PROPERTIES_FILE);
            props.load(input);
            if (props.get("useragent.version") != null) {
                this.userAgent = "java-sdk-" + props.get("useragent.version");
            }
        } catch (IOException e) {
            throw new RuntimeException("Properties file " + PROPERTIES_FILE + " is not found");
        }
    }

    private void checkParam(String region, String uid) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(region), "Region cannot be blank");
        Preconditions.checkArgument(region.matches(REGION_PATTERN), "Illegal region");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(uid), "Account ID cannot be blank");
        Preconditions.checkArgument(uid.matches(UID_PATTERN), "Illegal Account ID");
    }
    
    public Config(String region, String uid, String accessKeyID, String accessKeySecret, String securityToken,
            boolean isHttps) {
        this(region, uid, isHttps);
        
        Preconditions.checkArgument(!Strings.isNullOrEmpty(accessKeyID), "Access key cannot be blank");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(accessKeySecret), "Secret key cannot be blank");
        
        this.accessKeyID = accessKeyID;
        this.accessKeySecret = accessKeySecret;
        this.securityToken = securityToken;
    }

    /**
     * init with CredentialProvider for non-AK accessing (ECS instance)
     * 
     * @param region
     * @param uid
     * @param credsProvider
     * @param isHttps
     */
    public Config(String region, String uid, AlibabaCloudCredentialsProvider credsProvider, boolean isHttps) {
        this(region, uid, isHttps);
        
        this.credsProvider = credsProvider;
    }

    public String getUid() {
        return uid;
    }

    public Config setUid(String uid) {
        this.uid = uid;
        return this;
    }

    public String getAccountId() {
        return accountId;
    }

    public Config setAccoutId(String accountId) {
        this.accountId = accountId;
        return this;
    }

    public String getEndpoint() {
        return endpoint;
    }

    public Config setEndpoint(String endpoint) {
        this.endpoint = endpoint;
        return this;
    }

    public String getApiVersion() {
        return apiVersion;
    }

    public Config setApiVersion(String apiVersion) {
        this.apiVersion = apiVersion;
        return this;
    }

    public String getAccessKeyID() {
        return accessKeyID;
    }

    public Config setAccessKeyID(String accessKeyID) {
        this.accessKeyID = accessKeyID;
        return this;
    }

    public String getAccessKeySecret() {
        return accessKeySecret;
    }

    public Config setAccessKeySecret(String accessKeySecret) {
        this.accessKeySecret = accessKeySecret;
        return this;
    }

    public Boolean getDebug() {
        return isDebug;
    }

    public Config setDebug(Boolean debug) {
        this.isDebug = debug;
        return this;
    }

    public int getConnectTimeoutMillis() {
        return connectTimeoutMillis;
    }

    /**
     * Sets a specified timeout value, in milliseconds, to be used when opening a
     * communications link to the resource referenced by this URLConnection. If the
     * timeout expires before the connection can be established, a
     * java.net.SocketTimeoutException is raised. A timeout of zero is interpreted
     * as an infinite timeout.
     *
     * Also see
     * http://docs.oracle.com/javase/6/docs/api/java/net/URLConnection.html#setReadTimeout%28int%29
     * 
     * @return
     */
    public Config setConnectTimeoutMillis(int connectTimeoutMillis) {
        this.connectTimeoutMillis = connectTimeoutMillis;
        return this;
    }

    public int getConnectionRequestTimeoutMillis() {
        return connectionRequestTimeoutMillis;
    }

    public Config setConnectionRequestTimeoutMillis(int connectionRequestTimeoutMillis) {
        this.connectionRequestTimeoutMillis = connectionRequestTimeoutMillis;
        return this;
    }

    public int getReadTimeoutMillis() {
        return readTimeoutMillis;
    }

    /**
     * Sets the read timeout to a specified timeout, in milliseconds. A non-zero
     * value specifies the timeout when reading from Input stream when a connection
     * is established to a resource. If the timeout expires before there is data
     * available for read, a java.net.SocketTimeoutException is raised. A timeout of
     * zero is interpreted as an infinite timeout.
     *
     * Also see
     * http://docs.oracle.com/javase/6/docs/api/java/net/URLConnection.html#setReadTimeout%28int%29
     * 
     * @param readTimeoutMillis
     * @return
     */
    public Config setReadTimeoutMillis(int readTimeoutMillis) {
        this.readTimeoutMillis = readTimeoutMillis;
        return this;
    }

    public String getHost() {
        return host;
    }

    public Config setHost(String host) {
        this.host = host;
        return this;
    }

    public String getUserAgent() {
        return userAgent;
    }

    public String getSecurityToken() {
        return securityToken;
    }

    public Config setSecurityToken(String securityToken) {
        this.securityToken = securityToken;
        return this;
    }

    private String buildEndpoint(String region, String uid, boolean isHttps) {
        String endpoint = String.format(ENDPOINT_FMT, uid, region);
        String protocol = isHttps ? "https" : "http";
        return protocol + "://" + endpoint;
    }

    /**
     * refresh credentials if CredentialProvider set
     */
    public void refreshCredentials() {
        if (this.credsProvider == null)
            return;

        try {
            AlibabaCloudCredentials creds = this.credsProvider.getCredentials();
            this.accessKeyID = creds.getAccessKeyId();
            this.accessKeySecret = creds.getAccessKeySecret();

            if (creds instanceof BasicSessionCredentials) {
                this.securityToken = ((BasicSessionCredentials) creds).getSessionToken();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public AlibabaCloudCredentialsProvider getCredsProvider() {
        return credsProvider;
    }

    public void setCredsProvider(AlibabaCloudCredentialsProvider credsProvider) {
        this.credsProvider = credsProvider;
    }

    public int getThreadCount() {
        return threadCount;
    }

    public void setThreadCount(int threadCount) {
        this.threadCount = threadCount;
    }

    /**
     * Async client max connection count
     * @return Max connection count
     */
    public int getMaxConnectCount() {
        return maxConnectCount;
    }

    public void setMaxConnectCount(int maxConnectCount) {
        this.maxConnectCount = maxConnectCount;
    }

    /**
     * Async client max connection count per route
     * @return Max count per route
     */
    public int getMaxPerRoute() {
        return maxPerRoute;
    }

    public void setMaxPerRoute(int maxPerRoute) {
        this.maxPerRoute = maxPerRoute;
    }
}